spec/Classes.tex

```diff
@@ -663,73 +604,456 @@ \subsection{User-Defined Constructors}
+In the following code:
+\begin{chapel}
+class Point3D {
+  var x = 1.0;
+  var y = 1.0;
+  var z = 1.0;
+
+  proc init(x: real) {
+    this.x = x;
+    // compiler inserts "this.y = 1.0;"
+    this.z = y * 2.0;
+  }
+
+  proc init(x: real, y: real, z: real) {
+    this.x = x;
+    this.z = z;
+    this.y = y; // Error!
+  }
+}
+
+var A = new Point3D(1.0);
+var B = new Point3D(1.0, 2.0, 3.0);
+\end{chapel}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+\begin{chapeloutput}
+fieldsOutOfOrder.chpl:12: In initializer:
+fieldsOutOfOrder.chpl:17: error: Field "y" initialized out of order
+fieldsOutOfOrder.chpl:17: note: initialization of fields before .init() call must be in field declaration order
+\end{chapeloutput}
+
+The first initializer leverages the well-defined order of omitted field
+initialization to use the default value of field \chpl{y} in order to
+explicitly initialize field \chpl{z}.
+
+The second initializer initializes field \chpl{z} before field \chpl{y},
+causing a compile-time error to be issued.
+
+\end{chapelexample}
+
+\begin{rationale}
+Without this rule the compiler could insert default initialization for field
+\chpl{y} before \chpl{z} is explicitly initialized. The following statement
+would then be {\em assignment} to field \chpl{y}, despite appearing to be
+initialization. This subtle difference may be confusing and surprising, and
+is avoided by requiring fields to be initialized in field declaration order.
+\end{rationale}
+
+\subsubsection{The complete Method}
+\label{The_Complete_Method}
+\index{classes!initializers!user-defined!complete}
+
+As the initializer makes progress, the class instance is incrementally
+initialized. In order to prevent usage of uninitialized memory, there are
+restrictions on usage of the class instance before it is fully initialized:
+
+\begin{itemize}
+\item Methods may not be invoked on partially-initialized instances
+\item \chpl{this} may not be passed to functions while partially-initialized
+\end{itemize}
+
+These rules allow all methods and functions to assume that class instances
+have been initialized, provided their value is not \chpl{nil}.
+
+\begin{rationale}
+  The compiler could conceivably attempt to analyze methods and functions to
+  determine which fields are used, and selectively allow method calls on
+  partially-initialized class instances. A simpler rule, which is easier to
+  remember and implement, is to forbid method calls on partially-initialized
+  instances.
+\end{rationale}
+
+Methods may be called and \chpl{this} may be passed to functions only after the
+built-in \chpl{complete} method is invoked.
+
+\begin{rationale}
+Due to support for omitted field initialization, there is potential for
+confusion regarding the overall status of initialization. This confusion is
+addressed in the design by requiring \chpl{complete} to explicitly mark the
+transition between partially and fully initialized instances.
+\end{rationale}
```

Comment:
I would move this rationale to the beginning of the section

Reply:
I felt that it was better to introduce the context for the problem and the solution before rationalizing it.

Reply:
But this rationale is also introducing context.  At least, it seems a little out of place here at the bottom of the section

Reply:
Pretty sure I had asked Ben to add the rationale section later. On the spec itself, as a style guideline, I think it should say what the language design is and then, in a rationale section, say why we have that design.